h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1754,76 +1754,63 @@ private int getMovePriority(Chunk chunk) {
         if (move != null) {
-            long leftmostBlock = Long.MAX_VALUE;
-            for (Chunk chunk : move) {
-                leftmostBlock = Math.min(leftmostBlock, chunk.block);
-            }
-
             // this will ensure better recognition of the last chunk
             // in case of power failure, since we are going to move older chunks
             // to the end of the file
             writeStoreHeader();
             sync();
 
-            long originalFileSize = getFileLengthInUse();
-            for (Iterator<Chunk> iter = move.iterator(); iter.hasNext(); ) {
-                Chunk chunk = iter.next();
-                // while chunks destination do not overlap with their
-                // source location as a group, we can move them directly
-                // to the target destination
-                if (fileStore.predictAllocation(chunk.len) + chunk.len > leftmostBlock) {
-                    // overlap encountered, now rest of the chunks
-                    // are going to be moved in two steps - first to EOF,
-                    // and only then to target destination
-                    // this is to avoid possible corruption on abrupt shutdown
-                    break;
-                }
-                moveChunk(chunk, 0);
-                iter.remove();
+            long leftmostBlock = Long.MAX_VALUE;
+            for (Chunk chunk : move) {
+                leftmostBlock = Math.min(leftmostBlock, chunk.block);
             }
-
+            long sourceAreaLow = leftmostBlock * BLOCK_SIZE;
+            long originalFileSize = getFileLengthInUse();
             boolean movedToEOF = false;
             // we need to ensure that chunks moved within the following loop
             // do not overlap with space just released by chunks moved above,
             // if some of them happened to sit right at the end of used space,
             // hence the need to reserve this area
             for (Chunk chunk : move) {
-                moveChunk(chunk, originalFileSize);
-                movedToEOF = true;
+                moveChunk(chunk, sourceAreaLow, originalFileSize);
+                movedToEOF = movedToEOF || chunk.block >= originalFileSize;
             }
 
             long biggestFileSize = getFileLengthInUse();
-            Chunk chunkWithMetaToMove = null;
+            Chunk chunkToMove = null;
             if (movedToEOF) {
                 // update the metadata (store at the end of the file)
                 reuseSpace = false;
                 store(0, biggestFileSize);
                 sync();
 
-                chunkWithMetaToMove = lastChunk;
+                chunkToMove = lastChunk;
                 biggestFileSize = getFileLengthInUse();
 
                 // now re-use the empty space
                 reuseSpace = true;
                 for (Chunk c : move) {
-                    moveChunk(c, 0);
+                    if (c.block >= originalFileSize) {
```

Comment:
chunk.block(4096 bytes per block) can't be compared with originalFileSize(1 byte per position).